File: macos/Sources/NatKit/Models/SliceCommits.swift
import Foundation

/// The wire shape of `nat slice-diff --commits --project <id> --json <slice>`:
/// the branch's own history since the merge base, without diffing any of it —
/// what the sidebar's "All commits" dropdown lists (mirrors
/// `internal/cli/slicecommits.go`'s `commitsDoc`).
public struct SliceCommitsDoc: Codable, Equatable, Sendable {
    public let base: String
    public let branch: String
    public let commits: [SliceCommit]

    public init(base: String, branch: String, commits: [SliceCommit]) {
        self.base = base
        self.branch = branch
        self.commits = commits
    }
}

/// One commit of a branch's history: a hash, a subject line, who wrote it and
/// when — the four fields `git log` is asked for and nothing more (mirrors
/// `internal/git.Commit`/`internal/cli/slicecommits.go`'s `commitJSON`).
public struct SliceCommit: Codable, Equatable, Sendable, Identifiable {
    public let sha: String
    public let subject: String
    public let author: String
    public let date: Date

    public var id: String { sha }

    enum CodingKeys: String, CodingKey {
        case sha, subject, author, date
    }

    public init(sha: String, subject: String, author: String, date: Date) {
        self.sha = sha
        self.subject = subject
        self.author = author
        self.date = date
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        sha = try container.decode(String.self, forKey: .sha)
        subject = try container.decode(String.self, forKey: .subject)
        author = try container.decode(String.self, forKey: .author)
        let raw = try container.decode(String.self, forKey: .date)
        guard let parsed = PRDetail.parseGoTime(raw) else {
            throw DecodingError.dataCorruptedError(forKey: .date, in: container, debugDescription: "unreadable date: \(raw)")
        }
        date = parsed
    }

    /// The commit's own first eight hex characters — long enough to name one
    /// unambiguously in a list, mirrors `internal/cli/slicecommits.go`'s
    /// `shortSHA`.
    public var shortSHA: String {
        sha.count > 8 ? String(sha.prefix(8)) : sha
    }
}

File: internal/cli/slicecommits.go
package cli

import (
	"encoding/json"
	"fmt"
	"io"
	"strings"
	"time"

	"github.com/craigmjohnston/nat/internal/git"
)

// commitsDoc is the structured form of slice-diff --commits: the base and
// branch the commits are measured against, the same two fields the whole
// diff's own JSON carries, alongside the commits themselves.
type commitsDoc struct {
	Base    string       `json:"base"`
	Branch  string       `json:"branch"`
	Commits []commitJSON `json:"commits"`
}

// commitJSON is one commit, gh's own field names kept plain rather than
// nested: a hash, a subject line, who wrote it and when.
type commitJSON struct {
	SHA     string    `json:"sha"`
	Subject string    `json:"subject"`
	Author  string    `json:"author"`
	Date    time.Time `json:"date"`
}

// writeCommitsJSON encodes the branch's commits.
func writeCommitsJSON(out io.Writer, base, branch string, commits []git.Commit) error {
	doc := commitsDoc{Base: base, Branch: branch, Commits: make([]commitJSON, 0, len(commits))}
	for _, c := range commits {
		doc.Commits = append(doc.Commits, commitJSON{SHA: c.SHA, Subject: c.Subject, Author: c.Author, Date: c.Date})
	}
	enc := json.NewEncoder(out)
	enc.SetIndent("", "  ")
	return enc.Encode(doc)
}

// commitsMarkdown renders the commits as a list, newest first exactly as git
// log — and so [git.CLI.Commits] — already orders them.
func commitsMarkdown(base, branch string, commits []git.Commit) string {
	var b strings.Builder
	fmt.Fprintf(&b, "# Commits\n\n%s..%s\n\n", base, branch)
	if len(commits) == 0 {
		b.WriteString("_none_\n")
		return b.String()
	}
	for _, c := range commits {
		fmt.Fprintf(&b, "- %s %s — %s, %s\n", shortSHA(c.SHA), c.Subject, c.Author, c.Date.Format("2006-01-02"))
	}
	return b.String()
}

// shortSHA is the first eight characters of a commit's hash — long enough to
// name one unambiguously in a list, and short enough that the subject beside
// it is what a reader's eye lands on.
func shortSHA(sha string) string {
	if len(sha) > 8 {
		return sha[:8]
	}
	return sha
}
